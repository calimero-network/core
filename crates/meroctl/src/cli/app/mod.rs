use clap::{Parser, Subcommand};

use super::RootArgs;

mod install;
mod list;

#[derive(Parser, Debug)]
pub struct AppCommand {
    #[command(subcommand)]
    pub subcommand: AppSubCommands,
}

#[derive(Debug, Subcommand)]
pub enum AppSubCommands {
    Install(install::InstallCommand),
    #[command(alias = "ls")]
    List(list::ListCommand),
}

impl AppCommand {
    pub async fn run(self, args: RootArgs) -> eyre::Result<()> {
        match self.subcommand {
            AppSubCommands::Install(install) => install.run(args).await,
            AppSubCommands::List(list) => list.run(args).await,
        }
    }
}
