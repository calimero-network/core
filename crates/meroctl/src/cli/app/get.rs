use clap::{Parser, ValueEnum};
use eyre::{bail, Result as EyreResult};
use reqwest::Client;

use crate::cli::RootArgs;
use crate::common::RequestType::GET;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Parser, Debug)]
pub struct GetCommand {
    #[arg(long, short)]
    pub method: GetValues,

    #[arg(long, short)]
    pub app_id: String,
}
#[derive(ValueEnum, Debug, Clone)]
pub enum GetValues {
    Details,
}

impl GetCommand {
    pub async fn run(self, args: RootArgs) -> EyreResult<()> {
        let path = args.home.join(&args.node_name);

        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        };

        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file")
        };

        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let client = Client::new();

        let url = multiaddr_to_url(
            multiaddr,
            &format!("admin-api/dev/applications/{}", self.app_id),
        )?;

        let response = get_response(&client, url, None::<()>, &config.identity, GET).await?;

        if !response.status().is_success() {
            bail!("Request failed with status: {}", response.status())
        }

        println!("{}", response.text().await?);

        Ok(())
    }
}
