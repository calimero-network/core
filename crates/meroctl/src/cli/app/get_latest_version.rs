use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Clone, Parser, Debug)]
#[command(about = "Get the latest version of a package")]
pub struct GetLatestVersionCommand {
    #[arg(help = "Package name (e.g., com.example.myapp)")]
    pub package: String,
}

impl GetLatestVersionCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.get_latest_version(&self.package).await?;

        environment.output.write(&response);
        Ok(())
    }
}
