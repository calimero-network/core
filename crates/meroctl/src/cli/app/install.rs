use calimero_primitives::hash::Hash;
use calimero_server_primitives::admin::{
    InstallApplicationRequest, InstallApplicationResponse, InstallDevApplicationRequest,
};
use camino::Utf8PathBuf;
use clap::Parser;
use eyre::{bail, Result};
use reqwest::Client;
use semver::Version;
use tracing::info;
use url::Url;

use crate::cli::RootArgs;
use crate::common::RequestType::POST;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Debug, Parser)]
pub struct InstallCommand {
    /// Path to the application
    #[arg(long, short, conflicts_with = "url")]
    pub path: Option<Utf8PathBuf>,

    /// Url of the application
    #[clap(long, short, conflicts_with = "path", requires = "metadata")]
    pub url: Option<String>,

    #[clap(short, long, help = "Version of the application")]
    pub version: Option<Version>,

    #[clap(short, long, help = "Metadata for the application")]
    pub metadata: Option<Vec<u8>>,

    #[clap(long, help = "Hash of the application")]
    pub hash: Option<Hash>,
}

impl InstallCommand {
    pub async fn run(self, args: RootArgs) -> Result<()> {
        let path = args.home.join(&args.node_name);

        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        };

        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file")
        };

        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let client = Client::new();

        let mut is_dev_installation = false;

        let install_request = if let Some(app_path) = self.path {
            let install_dev_request = InstallDevApplicationRequest::new(
                app_path.canonicalize_utf8()?,
                self.version,
                self.metadata.unwrap_or_default(),
            );
            is_dev_installation = true;
            serde_json::to_value(install_dev_request)?
        } else if let Some(app_url) = self.url {
            let install_request = InstallApplicationRequest::new(
                Url::parse(&app_url)?,
                self.version,
                self.hash,
                self.metadata.unwrap_or_default(),
            );
            serde_json::to_value(install_request)?
        } else {
            bail!("Either path or url must be provided");
        };

        let install_url = if is_dev_installation {
            multiaddr_to_url(multiaddr, "admin-api/dev/install-dev-application")?
        } else {
            multiaddr_to_url(multiaddr, "admin-api/dev/install-application")?
        };

        let install_response = get_response(
            &client,
            install_url,
            Some(install_request),
            &config.identity,
            POST,
        )
        .await?;

        if !install_response.status().is_success() {
            let status = install_response.status();
            let error_text = install_response.text().await?;
            bail!(
                "Application installation failed with status: {}. Error: {}",
                status,
                error_text
            )
        }

        let body = install_response
            .json::<InstallApplicationResponse>()
            .await?;

        info!(
            "Application installed successfully. Application ID: {}",
            body.data.application_id
        );

        Ok(())
    }
}
