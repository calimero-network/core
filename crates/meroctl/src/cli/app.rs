use clap::{Parser, Subcommand};
use eyre::Result as EyreResult;

use super::RootArgs;
use crate::cli::app::get::GetCommand;
use crate::cli::app::install::InstallCommand;
use crate::cli::app::list::ListCommand;

mod get;
mod install;
mod list;

#[derive(Debug, Parser)]
pub struct AppCommand {
    #[command(subcommand)]
    pub subcommand: AppSubCommands,
}

#[derive(Debug, Subcommand)]
pub enum AppSubCommands {
    Get(GetCommand),
    Install(InstallCommand),
    #[command(alias = "ls")]
    List(ListCommand),
}

impl AppCommand {
    pub async fn run(self, args: RootArgs) -> EyreResult<()> {
        match self.subcommand {
            AppSubCommands::Get(get) => get.run(args).await,
            AppSubCommands::Install(install) => install.run(args).await,
            AppSubCommands::List(list) => list.run(args).await,
        }
    }
}
