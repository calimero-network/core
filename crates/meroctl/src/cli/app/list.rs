use calimero_server_primitives::admin::ListApplicationsResponse;
use clap::Parser;
use eyre::{bail, Result as EyreResult};
use reqwest::Client;

use crate::cli::RootArgs;
use crate::common::RequestType::GET;
use crate::common::{get_response, multiaddr_to_url};
use crate::config_file::ConfigFile;

#[derive(Debug, Parser)]
pub struct ListCommand;

impl ListCommand {
    pub async fn run(self, root_args: RootArgs) -> EyreResult<()> {
        let path = root_args.home.join(&root_args.node_name);
        if !ConfigFile::exists(&path) {
            bail!("Config file does not exist")
        }
        let Ok(config) = ConfigFile::load(&path) else {
            bail!("Failed to load config file");
        };
        let Some(multiaddr) = config.network.server.listen.first() else {
            bail!("No address.")
        };

        let url = multiaddr_to_url(multiaddr, "admin-api/dev/applications")?;
        let client = Client::new();
        let response = get_response(&client, url, None::<()>, &config.identity, GET).await?;

        if !response.status().is_success() {
            bail!("Request failed with status: {}", response.status())
        }

        let api_response: ListApplicationsResponse = response.json().await?;
        let app_list = api_response.data.apps;

        #[allow(clippy::print_stdout)]
        for app in app_list {
            println!("{}", app.id);
        }

        Ok(())
    }
}
