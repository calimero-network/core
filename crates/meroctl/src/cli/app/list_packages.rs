use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Copy, Clone, Parser, Debug)]
#[command(about = "List all packages")]
pub struct ListPackagesCommand;

impl ListPackagesCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.list_packages().await?;

        environment.output.write(&response);
        Ok(())
    }
}
