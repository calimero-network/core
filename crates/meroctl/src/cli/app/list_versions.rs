use clap::Parser;
use eyre::Result;

use crate::cli::Environment;

#[derive(Clone, Parser, Debug)]
#[command(about = "List versions of a package")]
pub struct ListVersionsCommand {
    #[arg(help = "Package name (e.g., com.example.myapp)")]
    pub package: String,
}

impl ListVersionsCommand {
    pub async fn run(self, environment: &mut Environment) -> Result<()> {
        let client = environment.client()?;

        let response = client.list_versions(&self.package).await?;

        environment.output.write(&response);
        Ok(())
    }
}
