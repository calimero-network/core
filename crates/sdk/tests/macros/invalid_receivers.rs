use calimero_sdk::app;

#[app::state]
struct MyType<'a>(&'a ());

#[app::logic]
impl<'a> MyType<'a> {
    pub fn method_00(self) {}
    pub fn method_01(self: Self) {}
    pub fn method_02(self: (Self)) {}
    pub fn method_03(self: (Self,)) {}
    pub fn method_04(mut self: Self) {}
    pub fn method_05(mut self: (Self)) {}
    pub fn method_06(self: &'a (Self,)) {}
    pub fn method_07(self: MyType<'a>) {}
    pub fn method_08(self: (MyType<'a>)) {}
    pub fn method_09(self: (MyType<'a>,)) {}
    pub fn method_10(mut self: MyType<'a>) {}
    pub fn method_11(mut self: (MyType<'a>)) {}
    pub fn method_12(mut self: (MyType<'a>,)) {}
    pub fn method_13(self: OtherType) {}
    pub fn method_14(self: (OtherType)) {}
    pub fn method_15(self: (OtherType,)) {}
    pub fn method_16(self: &OtherType) {}
    pub fn method_17(self: &(OtherType)) {}
    pub fn method_18(self: &(OtherType,)) {}
}

struct OtherType;

fn main() {}
