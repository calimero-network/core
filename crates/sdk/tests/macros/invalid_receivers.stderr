error: (calimero)> `self` must be passed by reference
 --> tests/macros/invalid_receivers.rs:8:22
  |
8 |     pub fn method_00(self) {}
  |                      ^^^^

error: (calimero)> `self` must be passed by reference
 --> tests/macros/invalid_receivers.rs:9:28
  |
9 |     pub fn method_01(self: Self) {}
  |                            ^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:10:29
   |
10 |     pub fn method_02(self: (Self)) {}
   |                             ^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:12:32
   |
12 |     pub fn method_04(mut self: Self) {}
   |                                ^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:13:33
   |
13 |     pub fn method_05(mut self: (Self)) {}
   |                                 ^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:15:28
   |
15 |     pub fn method_07(self: MyType<'a>) {}
   |                            ^^^^^^^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:16:29
   |
16 |     pub fn method_08(self: (MyType<'a>)) {}
   |                             ^^^^^^^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:18:32
   |
18 |     pub fn method_10(mut self: MyType<'a>) {}
   |                                ^^^^^^^^^^

error: (calimero)> `self` must be passed by reference
  --> tests/macros/invalid_receivers.rs:19:33
   |
19 |     pub fn method_11(mut self: (MyType<'a>)) {}
   |                                 ^^^^^^^^^^

error: (calimero)> expected `Self` or `MyType<'a>`
  --> tests/macros/invalid_receivers.rs:21:28
   |
21 |     pub fn method_13(self: OtherType) {}
   |                            ^^^^^^^^^

error: (calimero)> expected `Self` or `MyType<'a>`
  --> tests/macros/invalid_receivers.rs:22:29
   |
22 |     pub fn method_14(self: (OtherType)) {}
   |                             ^^^^^^^^^

error: (calimero)> expected `Self` or `MyType<'a>`
  --> tests/macros/invalid_receivers.rs:24:29
   |
24 |     pub fn method_16(self: &OtherType) {}
   |                             ^^^^^^^^^

error: (calimero)> expected `Self` or `MyType<'a>`
  --> tests/macros/invalid_receivers.rs:25:30
   |
25 |     pub fn method_17(self: &(OtherType)) {}
   |                              ^^^^^^^^^

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:10:28
   |
10 |     pub fn method_02(self: (Self)) {}
   |                            ^    ^
   |
   = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
   |
10 -     pub fn method_02(self: (Self)) {}
10 +     pub fn method_02(self: Self) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:13:32
   |
13 |     pub fn method_05(mut self: (Self)) {}
   |                                ^    ^
   |
help: remove these parentheses
   |
13 -     pub fn method_05(mut self: (Self)) {}
13 +     pub fn method_05(mut self: Self) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:16:28
   |
16 |     pub fn method_08(self: (MyType<'a>)) {}
   |                            ^          ^
   |
help: remove these parentheses
   |
16 -     pub fn method_08(self: (MyType<'a>)) {}
16 +     pub fn method_08(self: MyType<'a>) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:19:32
   |
19 |     pub fn method_11(mut self: (MyType<'a>)) {}
   |                                ^          ^
   |
help: remove these parentheses
   |
19 -     pub fn method_11(mut self: (MyType<'a>)) {}
19 +     pub fn method_11(mut self: MyType<'a>) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:22:28
   |
22 |     pub fn method_14(self: (OtherType)) {}
   |                            ^         ^
   |
help: remove these parentheses
   |
22 -     pub fn method_14(self: (OtherType)) {}
22 +     pub fn method_14(self: OtherType) {}
   |

warning: unnecessary parentheses around type
  --> tests/macros/invalid_receivers.rs:25:29
   |
25 |     pub fn method_17(self: &(OtherType)) {}
   |                             ^         ^
   |
help: remove these parentheses
   |
25 -     pub fn method_17(self: &(OtherType)) {}
25 +     pub fn method_17(self: &OtherType) {}
   |

error[E0307]: invalid `self` parameter type: (MyType<'a>,)
  --> tests/macros/invalid_receivers.rs:11:28
   |
11 |     pub fn method_03(self: (Self,)) {}
   |                            ^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: &'a (MyType<'a>,)
  --> tests/macros/invalid_receivers.rs:14:28
   |
14 |     pub fn method_06(self: &'a (Self,)) {}
   |                            ^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: (MyType<'a>,)
  --> tests/macros/invalid_receivers.rs:17:28
   |
17 |     pub fn method_09(self: (MyType<'a>,)) {}
   |                            ^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: (MyType<'a>,)
  --> tests/macros/invalid_receivers.rs:20:32
   |
20 |     pub fn method_12(mut self: (MyType<'a>,)) {}
   |                                ^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: OtherType
  --> tests/macros/invalid_receivers.rs:21:28
   |
21 |     pub fn method_13(self: OtherType) {}
   |                            ^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: OtherType
  --> tests/macros/invalid_receivers.rs:22:29
   |
22 |     pub fn method_14(self: (OtherType)) {}
   |                             ^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: (OtherType,)
  --> tests/macros/invalid_receivers.rs:23:28
   |
23 |     pub fn method_15(self: (OtherType,)) {}
   |                            ^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: &OtherType
  --> tests/macros/invalid_receivers.rs:24:28
   |
24 |     pub fn method_16(self: &OtherType) {}
   |                            ^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: &OtherType
  --> tests/macros/invalid_receivers.rs:25:28
   |
25 |     pub fn method_17(self: &(OtherType)) {}
   |                            ^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0307]: invalid `self` parameter type: &(OtherType,)
  --> tests/macros/invalid_receivers.rs:26:28
   |
26 |     pub fn method_18(self: &(OtherType,)) {}
   |                            ^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
